File: scripts/google-auth.ts
import * as http from 'http';
import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const SCOPES = [
  'https://www.googleapis.com/auth/gmail.readonly',
  'https://www.googleapis.com/auth/drive',
  'https://www.googleapis.com/auth/spreadsheets',
  'https://www.googleapis.com/auth/calendar.readonly'
];

const PORT = 8085;
const CREDENTIALS_PATH = path.join(__dirname, '..', 'gmail_credentials.json');
const TOKEN_PATH = path.join(__dirname, '..', 'credentials', 'token.json');

async function main() {
  // Load credentials
  const credentials = JSON.parse(fs.readFileSync(CREDENTIALS_PATH, 'utf-8'));
  const { client_id, client_secret } = credentials.installed;

  const redirect_uri = `http://localhost:${PORT}`;

  // Build auth URL
  const authUrl = new URL('https://accounts.google.com/o/oauth2/auth');
  authUrl.searchParams.set('client_id', client_id);
  authUrl.searchParams.set('redirect_uri', redirect_uri);
  authUrl.searchParams.set('response_type', 'code');
  authUrl.searchParams.set('scope', SCOPES.join(' '));
  authUrl.searchParams.set('access_type', 'offline');
  authUrl.searchParams.set('prompt', 'consent');

  console.log('\n=== GOOGLE OAUTH ===');
  console.log('\nOpen this URL in your browser:\n');
  console.log(authUrl.toString());
  console.log('\n\nWaiting for callback on port', PORT, '...\n');

  // Start server to catch callback
  const server = http.createServer(async (req, res) => {
    const url = new URL(req.url!, `http://localhost:${PORT}`);
    const code = url.searchParams.get('code');

    if (code) {
      console.log('Got authorization code. Exchanging for tokens...');

      try {
        // Exchange code for tokens
        const tokenResponse = await fetch('https://oauth2.googleapis.com/token', {
          method: 'POST',
          headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
          body: new URLSearchParams({
            code,
            client_id,
            client_secret,
            redirect_uri,
            grant_type: 'authorization_code'
          })
        });

        const tokens = await tokenResponse.json();

        if (tokens.error) {
          console.error('Token error:', tokens);
          res.writeHead(500);
          res.end('Error: ' + tokens.error_description);
          server.close();
          return;
        }

        // Add timestamp
        tokens.obtained_at = new Date().toISOString();

        // Save token
        fs.mkdirSync(path.dirname(TOKEN_PATH), { recursive: true });
        fs.writeFileSync(TOKEN_PATH, JSON.stringify(tokens, null, 2));

        console.log('\nToken saved to:', TOKEN_PATH);
        console.log('Access token:', tokens.access_token?.substring(0, 30) + '...');
        console.log('Refresh token:', tokens.refresh_token ? 'Present' : 'Missing');
        console.log('Expires in:', tokens.expires_in, 'seconds');

        res.writeHead(200, { 'Content-Type': 'text/html' });
        res.end('<h1>Success!</h1><p>You can close this window.</p>');

        server.close();
        process.exit(0);

      } catch (err) {
        console.error('Exchange error:', err);
        res.writeHead(500);
        res.end('Error exchanging code');
        server.close();
      }
    } else {
      res.writeHead(400);
      res.end('No code received');
    }
  });

  server.listen(PORT, () => {
    console.log(`Server listening on port ${PORT}`);
  });
}

main().catch(console.error);

File: scripts/google_auth.py
#!/usr/bin/env python3
"""
Google OAuth Setup for Frankie
Run this once to authorize Google Sheets access.
"""

import os
import json
from pathlib import Path
from http.server import HTTPServer, BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import webbrowser
import threading

# Check for required packages
try:
    from google_auth_oauthlib.flow import Flow
    from google.oauth2.credentials import Credentials
    from google.auth.transport.requests import Request
    from googleapiclient.discovery import build
except ImportError:
    print("Missing packages. Run:")
    print("  pip install google-auth-oauthlib google-auth-httplib2 google-api-python-client")
    exit(1)

SCOPES = [
    'https://www.googleapis.com/auth/spreadsheets',
    'https://www.googleapis.com/auth/drive.readonly'
]

CREDENTIALS_DIR = Path(__file__).parent.parent / 'credentials'
CLIENT_SECRET = CREDENTIALS_DIR / 'google_oauth.json'
TOKEN_FILE = CREDENTIALS_DIR / 'token.json'

# Store the auth code from callback
auth_code = None

class OAuthHandler(BaseHTTPRequestHandler):
    """Handle the OAuth callback."""

    def do_GET(self):
        global auth_code
        query = parse_qs(urlparse(self.path).query)

        if 'code' in query:
            auth_code = query['code'][0]
            self.send_response(200)
            self.send_header('Content-type', 'text/html')
            self.end_headers()
            self.wfile.write(b'''
                <html><body style="font-family: Arial; text-align: center; padding: 50px;">
                <h1>Authorization Successful!</h1>
                <p>You can close this window and return to the terminal.</p>
                </body></html>
            ''')
        else:
            self.send_response(400)
            self.send_header('Content-type', 'text/html')
            self.end_headers()
            error = query.get('error', ['Unknown error'])[0]
            self.wfile.write(f'<html><body><h1>Error: {error}</h1></body></html>'.encode())

    def log_message(self, format, *args):
        pass  # Suppress HTTP logs

def authenticate():
    """Run OAuth flow and save token."""
    global auth_code
    creds = None

    # Load existing token if available
    if TOKEN_FILE.exists():
        creds = Credentials.from_authorized_user_file(str(TOKEN_FILE), SCOPES)

    # Refresh or get new credentials
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            print("Refreshing expired token...")
            creds.refresh(Request())
        else:
            print("Starting OAuth flow...\n")

            # Use localhost redirect (no trailing slash — must match Google Console exactly)
            redirect_uri = 'http://localhost:8085'

            flow = Flow.from_client_secrets_file(
                str(CLIENT_SECRET),
                scopes=SCOPES,
                redirect_uri=redirect_uri
            )

            auth_url, _ = flow.authorization_url(
                access_type='offline',
                include_granted_scopes='true',
                prompt='consent'
            )

            print("=" * 60)
            print("STEP 1: Copy this URL and open it in your browser:\n")
            print(auth_url)
            print("\n" + "=" * 60)
            print("\nSTEP 2: Sign in with humanledai@gmail.com")
            print("STEP 3: After approval, you'll be redirected.")
            print("\nWaiting for authorization...\n")

            # Start local server to catch the callback
            server = HTTPServer(('localhost', 8085), OAuthHandler)
            server.timeout = 300  # 5 minute timeout

            # Try to open browser (may not work in WSL)
            try:
                webbrowser.open(auth_url)
            except:
                pass

            # Wait for callback
            while auth_code is None:
                server.handle_request()

            server.server_close()

            # Exchange code for token
            flow.fetch_token(code=auth_code)
            creds = flow.credentials

        # Save token for future use
        with open(TOKEN_FILE, 'w') as f:
            f.write(creds.to_json())
        print(f"\nToken saved to {TOKEN_FILE}")

    return creds

def test_sheets_access():
    """Test access to the tracking spreadsheet."""
    creds = authenticate()

    # Load sheet ID from env
    from dotenv import load_dotenv
    load_dotenv(CREDENTIALS_DIR / '.env')
    sheet_id = os.getenv('GOOGLE_SHEET_ID', '1_RfKelGdoc2teeAluqKJ1aYhXLBT-IwBCLeCYDy_CHI')

    try:
        service = build('sheets', 'v4', credentials=creds)
        sheet = service.spreadsheets()

        # Get spreadsheet metadata
        result = sheet.get(spreadsheetId=sheet_id).execute()
        print(f"\nConnected to: {result['properties']['title']}")
        print(f"Sheets: {[s['properties']['title'] for s in result['sheets']]}")
        print("\nGoogle Sheets access working!")
        return True
    except Exception as e:
        print(f"Error accessing sheet: {e}")
        return False

if __name__ == '__main__':
    print("=== Google OAuth Setup ===\n")
    test_sheets_access()
